sql/stripe/funnel/view.sql

```diff
@@ -0,0 +1,96 @@
+CREATE OR REPLACE VIEW
+  funnel
+AS
+WITH subscriptions AS (
+  SELECT
+    plan,
+    status = "incomplete_expired" AS incomplete_expired,
+    cancel_at_period_end
+    OR "cancelled_for_customer_at" IN (SELECT key FROM UNNEST(metadata)) AS cancelled_for_customer,
+    -- count subscriptions from the day after they start to the day after they end
+    -- TODO: decide whether to count users at the end of the day or the beginning of the day
+    DATE_ADD(DATE(start_date), INTERVAL 1 DAY) AS start_date,
+    -- use day after ended_at to count cancelled and renew failed
+    DATE_ADD(DATE(ended_at), INTERVAL 1 DAY) AS end_date,
+    trial_end,
+    MIN(DATE_ADD(DATE(start_date), INTERVAL 1 DAY)) OVER (
+      PARTITION BY
+        customer
+    ) AS customer_start_date,
+    customer,
+    -- TODO: join on guardian logs and cloudSQL for mozilla vpn funnel information (attribution,
+    -- client running, vpn active, waitlist)
+  FROM
+    stripe_derived.subscriptions_v1
+  WHERE
+    status != "incomplete"
+),
+counts AS (
+  SELECT
+    `date`,
+    plan,
+    CASE
+    WHEN
+      incomplete_expired
+    THEN
+      "failed setup"
+    WHEN
+      TIMESTAMP(`date`) < trial_end
+    THEN
+      "trialing"
+    -- TODO decide whether subscriptions that go through a trial are ever counted as "new"
```

Comment:
I think we should count trials with phases "new trial" (on the day the trial starts), "trialing", "trial converted" (on the day the paid subscription starts), and "trial cancelled" (on the day the trial ended without becoming paid)

Reply:
so essentially 2

Reply:
Ah - one wrinkle this might have highlighted! I think the count of active subscriptions for e.g. 2020-09-01 should include all subscriptions that were active at 2020-09-01 00:00:00 UTC, regardless of whether they are "new" (started immediately before this moment) or whether they were cancelled immediately after this moment. So a subscription can be "active" _and_ another state on a given date.

Basically, we're querying two quite-separate quantities:

1. "how many subscriptions were active/in a trial at each of these _moments_"
2. "how many subscriptions made transition _x_ during each of these _time periods_"

In (1) we're counting users in a state at midnight each day, and in (2) we're counting how many users change between sets of states during the time between midnights.

Reply:
going to split this into two queries, one for active subs by date, and a separate one for transitions